### PR TITLE
RFC: Remove proxy-deps

### PR DESCRIPTION
**Pull request by olix0r**
## Problem

Currently, the proxy-deps docker image is only of marginal utility: Though we build the project (twice!), the produced artifacts aren't used by the proxy image. This means that when Cargo.lock changes, we waste as much as 3x build time; and when a proxy source changes, we rebuild all dependencies and not just the proxy.

## Solution

In order to fix this, we can use a multi-stage dockerfile that builds successive components of the proxy project to optimize for cache. This is done by mocking unneeded sources to trigger dependency build.

### gRPC & build dependencies

When building the proxy, build-time dependencies are required to generate  gRPC bindings. These dependencies could not be cached without moving them into a dedicated crate. This seems like a good idea, more generally.

conduit-grpc's _arbitrary_ feature flag is necessary to implement quickcheck's `Arbitrary` for protobuf bindings to be used by tests in conduit-proxy. These implementations are protected by a feature flag so that Conduit doesn't take a runtime dependency on quickcheck.

Furthermore, in order to create a new _conduit-grpc_ crate, the _convert_ module also has to be moved into its on crate.

While we were making all of these little crates, it seemed like better hygiene just to stuff all of the rust libraries in the `proxy/` directory so they aren't exposed on the root of the source tree.

### Dockerfile build-args

While doing this, I additionally changed the behavior of `proxy/Dockerfile`:
- Added `RUST_IMAGE` build-arg, to set the build image.
- Added `RUNTIME_IMAGE` build-arg to set the final runtime base image.
- Proxy tests are now run when building docker images, when the `SKIP_TESTS` build-arg is not set.
- The `RELEASE` build-arg has been replaced by the `DEBUG` build-arg. Release builds were aleady the default behavior, but now it is clearer in the dockerfile.

### `bin/docker-build-proxy`

* No longer attempts to build the base image. This was a legacy of a time before images were publicly available.
* Accepts command-line arguments that will be passed directly to Docker. For example:

```
bin/docker-build-proxy --build-arg DEBUG=1 --build-arg SKIP_TESTS=1 --no-cache
```

## Testing

Running the following with Docker for Mac:

```sh
# Trigger a full dependency update and rebuild.
echo "# $(date )" >>Cargo.lock
time bin/docker-build-proxy --target=bin --build-arg "DEBUG=$DEBUG" --build-arg "SKIP_TESTS=1"

# Simulate a proxy-only edit and rebuild.
echo >>proxy/src/main.rs
time bin/docker-build-proxy --target=bin --build-arg "DEBUG=$DEBUG" --build-arg "SKIP_TESTS=1"

# Measure only the added time of testing.
time bin/docker-build-proxy --target=bin --build-arg "DEBUG=$DEBUG"
```

|  | Full | Edit | Tests |
| -------|---------|--------|---------|
| `DEBUG=1` | 3:59.69 | 0:38.331 | 1:23.55 |
| `DEBUG=`  |  7:40.17 | 2:16.99 | 4:55.98 |

### Previously

```sh
echo "# $(date )" >>Cargo.lock
bin/update-proxy-deps-shas
time bin/docker-build-proxy

# Simulate a proxy-only edit and rebuild.
echo >>proxy/src/main.rs
time bin/docker-build-proxy
```

| Full        | Edit |
|---------|--------|
| 17:18.59 | 14:02.35 |

What's more, docker-build-proxy _can't_ produce artifacts with debug symbols (due to a shell expansion glitch).

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -123,10 +120,10 @@ jobs:
         - |
           export CONDUIT_TAG=$(. bin/_tag.sh ; clean_head_root_tag)
           echo "CONDUIT_TAG=${CONDUIT_TAG}"
-        - export PROXY_RELEASE=1 BUILD_DEBUG=1 DOCKER_TRACE=1
+        - export BUILD_DEBUG=1 DOCKER_TRACE=1
 
       script:
-        - bin/docker-build
+        - SKIP_TESTS=1 bin/docker-build
```
Comment (briansmith) [inline .travis.yml]: Why are we skipping the tests here?

File: proxy/Dockerfile [hunk 2]
```diff
@@ -2,25 +2,110 @@
...
+#
+# Mocks out all local code and fetch external dependencies to ensure that external sources
+# are cached.
+FROM $RUST_IMAGE as fetch
+WORKDIR /usr/src/conduit
+COPY proxy/.empty.rs proxy/src/lib.rs
+COPY proxy/.empty.rs proxy/convert/src/lib.rs
+COPY proxy/.empty.rs proxy/futures-mpsc-lossy/src/lib.rs
+COPY proxy/.empty.rs proxy/grpc/src/lib.rs
+COPY proxy/.empty.rs proxy/tower-router/src/lib.rs
```
Comment (briansmith) [inline proxy/Dockerfile]: .empty.rs is confusing and distracting. Does `touch` not work here? If not, let's document why we're doing it this way and also include a note about it inside the .empty.rs file itself.

File: proxy/Dockerfile [hunk 3]
```diff
@@ -2,25 +2,110 @@
...
+RUN if [ -n "$DEBUG" ]; \
+    then cargo build --frozen ; \
+    else cargo build --frozen --release ; \
+    fi
+#RUN if [ -n "$SKIP_TESTS" ]; \
+#    then echo "tests skipped" ; \
+#    elif [ -n "$DEBUG" ]; \
+#    then cargo test --frozen ; \
+#    else cargo test --frozen --release ; \
+#    fi
```
Comment (briansmith) [inline proxy/Dockerfile]: Let's remove this commented-out stuff.

File: proxy/Dockerfile [hunk 4]
```diff
@@ -2,25 +2,110 @@
...
+RUN if [ -n "$DEBUG" ]; \
+    then cargo build -p conduit-grpc --features=arbitrary --frozen ; \
+    else cargo build -p conduit-grpc --features=arbitrary --frozen --release ; \
+    fi
+#RUN if [ -n "$SKIP_TESTS" ]; \
+#    then echo "tests skipped" ; \
+#    elif [ -n "$DEBUG" ]; \
+#    then cargo test -p conduit-grpc --features=arbitrary --frozen ; \
+#    else cargo test -p conduit-grpc --features=arbitrary --frozen --release ; \
+#    fi
```
Comment (briansmith) [inline proxy/Dockerfile]: Let's remove the commented-out stuff.

File: proxy/grpc/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,29 @@
+[package]
```
Comment (briansmith) [inline proxy/grpc/Cargo.toml]: 1. It would have been easier to review this if the factoring out of the generated gRPC bindings was done in a separate PR.

2. I think `api` or `proxy-api` is a better name for this than `conduit-grpc`.

File: proxy/src/control/pb.rs [hunk 6]
```diff
@@ -232,258 +175,3 @@ impl<'a> TryFrom<&'a Event> for common::TapEvent {
         Ok(tap_ev)
     }
 }
-
-impl<'a> TryFrom<&'a common::http_method::Type> for http::Method {
```
Comment (briansmith) [inline proxy/src/control/pb.rs]: Why is it suddenly OK to delete all of this? I'm not saying it isn't OK, but it's not clear what this refactoring has done to make it OK to delete all this. Is there any negative effect of deleting all this?

File: .travis.yml (hunk 1 above)
Comment (olix0r) [reply .travis.yml]: the tests have already been run

File: proxy/src/control/pb.rs (hunk 6 above)
Comment (olix0r) [reply proxy/src/control/pb.rs]: it's been moved with the grpc bindings

File: bin/docker-build [hunk 7]
```diff
@@ -9,6 +9,9 @@ fi
 
 bin/docker-build-controller
 bin/docker-build-web
-bin/docker-build-proxy
 bin/docker-build-proxy-init
 bin/docker-build-cli
+
+bin/docker-build-proxy \
+    --build-arg="SKIP_TESTS=${SKIP_TESTS:-}" \
+    --build-arg="DEBUG=${PROXY_DEBUG:-}"
```
Comment (siggy) [inline bin/docker-build]: can this all be `PROXY_DEBUG` ?

**Comment by olix0r** [comment]
going to discard this in favor of two smaller PRs

File: proxy/Dockerfile [hunk 8]
```diff
@@ -2,25 +2,110 @@
...
+COPY --from=grpc /usr/src/conduit/target ./target
+ARG DEBUG
+RUN if [ -n "$DEBUG" ]; \
+    then cargo build -p conduit-proxy --bin conduit-proxy --frozen ; \
+    else cargo build -p conduit-proxy --bin conduit-proxy --frozen --release ; \
+    fi
+ARG SKIP_TESTS
+RUN if [ -n "$SKIP_TESTS" ]; \
+    then echo "tests skipped" ; \
+    elif [ -n "$DEBUG" ]; \
```
[the remark this answers is not in the document]
Comment (olix0r) [reply proxy/Dockerfile]: not quite:

1. if SKIP_TESTS is set, just echo,
2. otherwise, if debug is set, test the debug build
3. otherwise, test the release build

I really think that tests should be run by default when producing artifacts.

File: proxy/Dockerfile [hunk 9]
```diff
@@ -2,25 +2,110 @@
 #
 # Builds a slim runtime image with the conduit-proxy binary.
 
-## Build the rust proxy into a binary.
+ARG RUST_IMAGE=rust:1.23.0
+
+ARG RUNTIME_IMAGE=gcr.io/runconduit/base:2017-10-30.01
```
[the remark this answers is not in the document]
Comment (olix0r) [reply proxy/Dockerfile]: Adding them later triggers the entire cache to be invalidated. I find them to be very useful for testing. They can safely be treated as if they're not there.
